### ISSUE TITLE
Provide variable bounds?

### ISSUE THREAD

**Issue by Yurlungur** [lanl/singularity-eos#69]
In a downstream code, I found I wanted to inspect the variable bounds in a tabular eos in an eos-independent way... in other words, I wanted meaningful bounds for even equations of state with no meaningful bounds.

I wonder if we should expose bounds for all EOS's? For those without bounds, we could use IEEE infinity? What do people think? @dholladay00 @jhp-lanl @annapiegra @brryan @chadmeyer particularly interested in your opinions on this.

**Comment by chadmeyer** [comment]
This seems like a useful addition to the analytic models. There are definitely some relevant bounds in some of the models.

**Comment by dholladay00** [comment]
Is this easy to implement in an EOS agnostic way? Do bounds exist for all relevant variables or will we internally get bounds for independent variables and then do lookups for dependent variables? My preference would be for fewer functions rather than more, kind of like the `getvaluesatreferencestate` thing.

**Comment by chadmeyer** [comment]
What bounds are we talking about, to make this discussion a bit more concrete?  Take, for instance, ideal gas.  Density is bounded at zero and infinity, sie at zero and infinity, pressure at zero and infinity, temperature at zero and infinity.  If we look at something like gruneisen form, our energies and pressures could be negative.  Some models might basically impose a minimum temperature (I think there might be something like that semi-included in one of the Davis EOSs, though maybe it's just more like a cold curve).

If it's *global* bounds we're talking about; most of these models will be simple to do, if it is otherwise constrained bounds (like temperature/pressure range given a particular density) then some cases might require more thought, but should still be doable.

The other relevant question to API design is what the intended use case(s) are.  Would something like this be useful in a PTE solver for instance, and how would it be used there?

**Comment by Yurlungur** [comment]
> The other relevant question to API design is what the intended use case(s) are. Would something like this be useful in a PTE solver for instance, and how would it be used there?

This is actually the context I found a need for this in. I wanted to do a one-dimensional root find for an astro application, and wanted enforce-able bounds. I think solvers are probably what should drive the design. 

I wonder if we should provide global and local bounds both... or just start with global to keep it simple and see if that's sufficient.

**Comment by chadmeyer** [comment]
The other question is bounds of which variables.  If you're thinking of a 1D root find, then `std::pair<Real,Real> range_of_x(some_variable)` seems most appropriate.  I suspect that for most analytic EOSs, this will be trivial (either 0:inf or -inf:inf), and there will just be a couple of situations where there is something like a cold curve's boundary on one side, but for most of the relevant cases, the information is already available to do that sort of thing.

**Comment by Yurlungur** [comment]
Agreed. On the other hand, I think there's a need for multi-D root finds too, a la PTE solvers. So I think it's better to design for that use case.

That said, for global bounds, multiple calls to `range_of_x` is fine. It's the local bounds where things get a bit gross.

**Comment by chadmeyer** [comment]
I suspect that a full 2D solve in PTE would benefit from `range_x(variable)` type calls, but I think we might have to actually sketch out how we would implement that in the first place before we built something for it.  The global bounds `range_x()` call is probably (practically at least) a constexpr for most analytic models.

**Comment by Yurlungur** [comment]
Yep. I will try to carve out some time and sketch something out in a PR.

**Comment by jhp-lanl** [comment]
What do you mean by "local" vs "global" bounds?

**Comment by Yurlungur** [comment]
Local meaning there might be some range of valid energies at a given density, for example.

Global meaning energies that might not be valid everywhere but are valid somewhere.

**Comment by jhp-lanl** [comment]
> Local meaning there might be some range of valid energies at a given density, for example.
> 
> Global meaning energies that might not be valid everywhere but are valid somewhere.

That makes sense. So is it accurate to say that local bounds would be pairs of functions of the other independent variable while global bounds are scalar pairs.

**Comment by Yurlungur** [comment]
Exactly.

**Comment by Yurlungur** [comment]
Partially resolved by #109. But we should probably leave this open as I think the discussion should probably be on-going.
